modules/porous_flow/src/utils/PorousFlowVanGenuchten.C

```diff
@@ -609,4 +609,268 @@ d2saturationHys(Real pc,
+                         Real y1p)
+{
+  if (sl < slr) // by the definition of slr, always return 0
+    return 0.0;
+  const Real sl_bar = (sl - slr) / (1.0 - slr); // effective saturation
+  const Real sl_bar_prime = 1.0 / (1.0 - slr);
+  // a and b are useful parameters.  Define b along the drying curve initially, and
+  // modify a and b appropriately if the wetting result is required
+  Real a = 0;
+  Real a_prime = 0.0;
+  Real b = 0;
+  Real b_prime = 0.0;
+  if (sgrdel == 0.0 || sl <= sldel) // along the drying curve
+  {
+    const Real c = std::pow(sl_bar, 1.0 / m);
+    const Real dc_dsbar = c / m / sl_bar;
+    a = std::pow(1.0 - c, m);
+    const Real da_dsbar = -m * a / (1.0 - c) * dc_dsbar;
+    a_prime = da_dsbar * sl_bar_prime;
+  }
+  else // along the wetting curve
+  {
+    // In most cases, use sldel and sgrdel as provided to this function.  However, because "there is
+    // no hysteresis along the extension" according to p6 of Doughty2008, if the turning point is
+    // less than slr, then use the expressions for the case when the turning point was slr
+    const Real my_sldel = (sldel < slr) ? slr : sldel;
+    const Real my_sgrdel = (sldel < slr) ? sgrmax : sgrdel;
+    if (sl >= 1.0 - 0.5 * my_sgrdel)
+    {
+      // follow the drying curve.  The parameter b has already been defined.  It
+      // is important for initialization of the curic that the above condition is >= and not >
+      const Real c = std::pow(sl_bar, 1.0 / m);
+      const Real dc_dsbar = c / m / sl_bar;
+      a = std::pow(1.0 - c, m);
+      const Real da_dsbar = -m * a / (1.0 - c) * dc_dsbar;
+      a_prime = da_dsbar * sl_bar_prime;
+    }
+    else if (sl > upper_liquid_param * (1.0 - my_sgrdel))
+    {
+      // follow the cubic modification of the wetting curve.  Immediately exit from this function by
+      // returning the cubic result
+      return dcubic(
+          sl, upper_liquid_param * (1.0 - my_sgrdel), y0, y0p, 1.0 - 0.5 * my_sgrdel, y1, y1p);
+    }
+    else
+    {
+      // standard case of wetting curve outside the cubic-modification and drying-curve regions
+      const Real sl_bar_del = (my_sldel - slr) / (1.0 - slr);
+      const Real s_gt_bar =
+          my_sgrdel * (sl - my_sldel) / (1.0 - slr) / (1.0 - my_sldel - my_sgrdel);
+      const Real s_gt_bar_prime = my_sgrdel / (1.0 - slr) / (1.0 - my_sldel - my_sgrdel);
+      const Real c = std::pow(sl_bar + s_gt_bar, 1.0 / m);
+      const Real c_prime = c / m / (sl_bar + s_gt_bar) * (sl_bar_prime + s_gt_bar_prime);
+      a = (1 - s_gt_bar / (1.0 - sl_bar_del)) * std::pow(1.0 - c, m);
+      a_prime =
+          -s_gt_bar_prime / (1.0 - sl_bar_del) * std::pow(1.0 - c, m) - m * a / (1.0 - c) * c_prime;
+      b = s_gt_bar / (1.0 - sl_bar_del) * std::pow(1.0 - std::pow(sl_bar_del, 1.0 / m), m);
+      b_prime = s_gt_bar_prime * b / s_gt_bar;
+    }
+  }
+  const Real kr = std::sqrt(sl_bar) * Utility::pow<2>(1.0 - a - b);
+  return 0.5 * kr / sl_bar * sl_bar_prime -
```

Comment:
Early return for `sl = slr` would avoid this blowing up when `sl_bar=0`.

Reply:
oooh, that's mean.  Yea, i agree.